File: nodejs/src/calc-witness.js
const fs = require("fs");
const { calc_witness } = require("witnesscalc");

function parseArgs() {
  const args = process.argv.slice(2);
  if (args.length !== 3) {
    console.error(
      `Usage: ${process.argv[0]} <graph.bin> <inputs.json> <witness.wtns>`,
    );
    process.exit(1);
  }

  const [graphFile, inputsFile, witnessFile] = args;

  if (!fs.existsSync(graphFile)) {
    console.error(`Graph file not found: ${graphFile}`);
    process.exit(1);
  }

  if (!fs.existsSync(inputsFile)) {
    console.error(`Inputs file not found: ${inputsFile}`);
    process.exit(1);
  }

  return { graphFile, inputsFile, witnessFile };
}

async function main() {
  const { graphFile, inputsFile, witnessFile } = parseArgs();

  const inputsData = await fs.promises.readFile(inputsFile, "utf-8");
  const graphData = await fs.promises.readFile(graphFile);
  const graphDataUint8Array = new Uint8Array(graphData.buffer, graphData.byteOffset, graphData.byteLength);

  const wtnsBytes = calc_witness(inputsData, graphDataUint8Array);

  await fs.promises.writeFile(witnessFile, wtnsBytes);
  console.log(`Witness saved to ${witnessFile}`);
}

main().catch((err) => {
  console.error(err);
  process.exit(1);
});

File: src/bin/calc-witness.rs
use circom_witnesscalc::{calc_witness_flow};
use std::env;
use std::fs::File;
use std::io::Write;

struct Args {
    graph_file: String,
    inputs_file: String,
    witness_file: String,
}

fn parse_args() -> Args {
    let args: Vec<String> = env::args().collect();
    if args.len() != 4 {
        eprintln!(
            "Usage: {} <graph.bin> <inputs.json> <witness.wtns>",
            args[0]
        );
        std::process::exit(1);
    }

    Args {
        graph_file: args[1].clone(),
        inputs_file: args[2].clone(),
        witness_file: args[3].clone(),
    }
}

fn main() {
    let args = parse_args();
    let inputs_data =
        std::fs::read_to_string(&args.inputs_file).expect("Failed to read input file");
    let graph_data = std::fs::read(&args.graph_file).expect("Failed to read graph file");

    let (wtns_bytes, _) = calc_witness_flow(&inputs_data, &graph_data);

    {
        let mut f = File::create(&args.witness_file).unwrap();
        f.write_all(&wtns_bytes).unwrap();
    }
    println!("witness saved to {}", &args.witness_file);
}

#[cfg(test)]
mod tests {
    #[test]
    fn test_ok() {
        println!("OK");
    }
}
